### PR TITLE
Add X509Ref::version

### PR DESCRIPTION
**Pull request by igorty**
Implements #1308.

### CODE REVIEW

File: openssl-sys/src/x509.rs [hunk 1]
```diff
@@ -225,6 +225,13 @@ extern "C" {
     pub fn X509_get_pubkey(x: *mut X509) -> *mut EVP_PKEY;
 
     pub fn X509_set_version(x: *mut X509, version: c_long) -> c_int;
+    /// Returns certificate version previously set through `X509_set_version` or the default which
+    /// is `0`.
+    ///
+    /// **Important.** The corresponding OpenSSL method is not&nbsp;expected to fail under its
+    /// correct implementation. Despite that, the method invocation passes through scopes which
+    /// _may_ result in error. Having that it is encouraged to validate returned value.
+    pub fn X509_get_version(x: *const X509) -> c_long;
```
Comment (sfackler) [inline openssl-sys/src/x509.rs:234]: There's no need to document function definitions in openssl-sys.

File: openssl/src/x509/mod.rs [hunk 2]
```diff
@@ -548,6 +589,20 @@ impl X509Ref {
...
+    ///
+    /// Note that `0` return value stands for version 1, `1`&nbsp;&ndash; for version 2 and so on.
+    ///
+    /// This corresponds to [`X509_get_version`].
+    ///
+    /// [`X509_get_version`]: https://www.openssl.org/docs/man1.1.1/man3/X509_get_version.html
+    pub fn version(&self) -> Result<i64, ErrorStack> {
+        // Covered with `x509_ref_version()`, `x509_ref_version_no_version_set()`,
+        // `x509_ref_version_incorrect_version_set()` tests
+        unsafe { cvt_long_n(ffi::X509_get_version(self.as_ptr())) }
```
Comment (sfackler) [inline openssl/src/x509/mod.rs:603]: The man page for X509_get_version does not claim it ever returns an error.

File: openssl-sys/src/x509.rs (hunk 1 above)
Comment (igorty) [reply openssl-sys/src/x509.rs:229]: Removed redundant doc.

File: openssl/src/x509/mod.rs (hunk 2 above)
Comment (igorty) [reply openssl/src/x509/mod.rs]: Removed error check.

**Comment by igorty** [comment]
It seems that tests on all platforms (except macos, which I haven't got available for run) [pass](https://app.circleci.com/pipelines/github/igorty/rust-openssl). There are still couple of concerns I'll raise as comments to the pull request changes.

File: openssl-sys/src/x509.rs [hunk 3]
```diff
@@ -225,6 +225,8 @@ extern "C" {
     pub fn X509_get_pubkey(x: *mut X509) -> *mut EVP_PKEY;
 
     pub fn X509_set_version(x: *mut X509, version: c_long) -> c_int;
+    #[cfg(ossl110)]
```
Comment (igorty) [inline openssl-sys/src/x509.rs:228]: It is unclear to me how comes that `X509_set_version` and `X509Builder::set_version` methods do not need `#[cfg(ossl110)]` constraint declared. OpenSSL docs for [v1.0.2](https://www.openssl.org/docs/man1.0.2/man3/) do not list a possibility to set X509 certificate version.

File: openssl/src/x509/mod.rs [hunk 4]
```diff
@@ -40,7 +40,63 @@ pub mod extension;
...
+    /// before a call to the tested method.
+    ///
+    /// _Note._ As for now this test is disabled since OpenSSL erroneously allows to set any number
+    /// as a certificate version. See corresponding [`issue`].
+    ///
+    /// [`issue`]: https://github.com/openssl/openssl/issues/12138
+    #[cfg(ossl110)]
+    #[test]
+    #[ignore]
+    fn x509_ref_version_incorrect_version_set() {
```
Comment (igorty) [inline openssl/src/x509/mod.rs:87]: Turns out there is a bug in OpenSSL `X509_set_version(X509 *x, long version)` function. It allows to set any number as described in the corresponding [issue](https://github.com/openssl/openssl/issues/12138). Probably it's worth to consider a temporary check in `X509Builder::set_version` and `X509ReqBuilder::set_version` while the mentioned bug is present?

File: openssl/src/x509/mod.rs [hunk 5]
```diff
@@ -548,6 +604,21 @@ impl X509Ref {
...
+    /// Note that `0` return value stands for version 1, `1`&nbsp;&ndash; for version 2 and so on.
+    ///
+    /// This corresponds to [`X509_get_version`].
+    ///
+    /// [`X509_get_version`]: https://www.openssl.org/docs/man1.1.1/man3/X509_get_version.html
+    #[cfg(ossl110)]
+    pub fn version(&self) -> i32 {
+        // Covered with `x509_ref_version()`, `x509_ref_version_no_version_set()`,
+        // `x509_ref_version_incorrect_version_set()` tests
+        unsafe { ffi::X509_get_version(self.as_ptr()) as i32 }
```
Comment (igorty) [inline openssl/src/x509/mod.rs:619]: In order `version()` method to work on both x32 and x64 OSs a cast to i32 is performed. Is this appropriate solution or should there be some kind of check performed in order to determine which `c_long` type definition was taken: from `libc-0.2.76/src/unix/linux_like/linux/gnu/b64/x86_64/not_x32.rs` or `libc-0.2.76/src/unix/linux_like/linux/gnu/b64/x86_64/x32.rs`?

File: openssl-sys/src/x509.rs (hunk 3 above)
Comment (sfackler) [reply openssl-sys/src/x509.rs:228]: OpenSSL's man pages do not cover every single function the library exposes.

File: openssl/src/x509/mod.rs (hunk 4 above)
Comment (sfackler) [reply openssl/src/x509/mod.rs:87]: I don't think it makes sense to write a test for a case that the underlying library does not handle.

File: openssl/src/x509/mod.rs (hunk 5 above)
Comment (sfackler) [reply openssl/src/x509/mod.rs:619]: This is fine.

File: openssl/src/x509/mod.rs (hunk 4 above)
Comment (igorty) [reply openssl/src/x509/mod.rs]: Removed this test.

File: openssl/src/x509/mod.rs [hunk 6]
```diff
@@ -548,6 +581,20 @@ impl X509Ref {
         }
     }
 
+    /// Returns certificate version. If this certificate has no explicit version set, it defaults to
+    /// version&nbsp;1.
```
Comment (sfackler) [inline openssl/src/x509/mod.rs:585]:
```suggestion
    /// version 1.
```

File: openssl/src/x509/mod.rs [hunk 7]
```diff
@@ -548,6 +581,20 @@ impl X509Ref {
         }
     }
 
+    /// Returns certificate version. If this certificate has no explicit version set, it defaults to
+    /// version 1.
+    ///
+    /// Note that `0` return value stands for version 1, `1`&nbsp;&ndash; for version 2 and so on.
```
Comment (sfackler) [inline openssl/src/x509/mod.rs:587]:
```suggestion
    /// Note that `0` return value stands for version 1, `1` for version 2 and so on.
```

**Pull request merged**